packages/ref-core/src/cmip_ref_core/pycmec/metric.py

```diff
@@ -0,0 +1,244 @@
+            self.root[dim_name] = dim_content
+
+    @classmethod
+    def merge_dimension(cls, metric_dim1, metric_dim2):
+        mdim1 = cls.model_validate(metric_dim1)
+        mdim2 = cls.model_validate(metric_dim2)
+
+        assert mdim1.root[MetricCV.JSON_STRUCTURE.value] == mdim2.root[MetricCV.JSON_STRUCTURE.value], (
+            "Error: JSON_STRUCTURES are not same"
+        )
+
+        merged_dim = {}
+        merged_dim = {MetricCV.JSON_STRUCTURE.value: mdim1.root[MetricCV.JSON_STRUCTURE.value]}
+
+        for dim in mdim1.root[MetricCV.JSON_STRUCTURE.value]:
+            merged_dim[dim] = mdim1.root[dim]
+
+        for dim in mdim2.root[MetricCV.JSON_STRUCTURE.value]:
+            for key in mdim2.root[dim].keys():
+                if key not in merged_dim[dim].keys():
+                    merged_dim[dim][key] = mdim2.root[dim][key]
+        return cls(merged_dim)
+
+
+class MetricResults(RootModel):
+    """
+    CMEC metric bundle RESULTS object
+    """
+
+    model_config = ConfigDict(strict=True)
+    root: dict[str, dict]
+
+    def __getitem__(self, key: str):
+        return self.root[key]
+
+    def __setitem__(self, key: str, value: dict):
+        self.root[key] = value
+
+    @classmethod
+    def _check_nested_dict_keys(cls, nested, metdims, level=0, keylevs=None):
+        if keylevs is None:
+            keylevs = {}
+
+        dim_name = metdims[MetricCV.JSON_STRUCTURE.value][level]
+
+        assert Counter(list(metdims[dim_name].keys())) == Counter(nested.keys()), "Error in dict"
+
+        for key, value in nested.items():
+            if level not in keylevs:
+                keylevs[level] = []
+            keylevs[level].append(key)
+            if isinstance(value, dict) and level < len(metdims[MetricCV.JSON_STRUCTURE.value]) - 1:
+                cls._check_nested_dict_keys(value, metdims, level + 1, keylevs)
+            elif isinstance(value, dict):
+                StrNumDict(value)
+
+        return keylevs
+
+    @model_validator(mode="after")
+    @classmethod
+    def validate_results(cls, rlt, info: ValidationInfo) -> dict:
+        if not isinstance(info.context, MetricDimensions):
+            print(
+                """To validate MetricResults object, MetricDimensions is needed,
+            please use model_validate(Results, context=MetricDimensions to instantiate"""
+            )
+            raise ValidationError
+        else:
+            results = rlt.root
+            metdims = info.context.root
+            keylevs = cls._check_nested_dict_keys(results, metdims, level=0, keylevs=None)
+
+        return rlt
+
+
+class StrNumDict(RootModel):
+    model_config = ConfigDict(strict=True)
+    root: dict[str, float]
+
+
+class CMECMetric(BaseModel):
+    """
+    CMEC metric bundle object
+    """
+
+    model_config = ConfigDict(strict=True)
+
+    SCHEMA: Optional[MetricSchema] = None
+    DIMENSIONS: MetricDimensions
+    RESULTS: dict
+    PROVENANCE: Optional[dict] = None
```

Comment:
We only support python >=3.10 so this can be written as:

```suggestion
    PROVENANCE: dict[str, Any] | None = None
```

Ruff will automatically rewrite these if you run `make ruff-fixes` or if you install the pre-commit hooks using `make virtual-environment` this will happen automatically when you commit.

The `dict` isn't a valid type hint by itself to mypy. It is a generic and needs to be provided with the type of the keys and values even if they are both Any.

Reply:
Thanks. yes. I have fixed these to meet the mypy requirement